tests/acct_mgt_test.py

```diff
@@ -0,0 +1,353 @@
+
+          auth_opts = ["-E","./client_cert/acct-mgt-2.crt", "-key", "./client_cert/acct-mgt-2.key"]
+          auth_opts = ["-cert", r"acct-mgt-2",]
+
+ Initial test to confirm that something is working
+    curl -kv https://acct-mgt.apps.cnv.massopen.cloud/projects/acct-mgt
+    curl -u <user>:<password> -kv https://acct-mgt.apps.cnv.massopen.cloud/projects/acct-mgt
+
+  -- testing with no authentication:
+     python3 -m pytest amt_test*.py --amurl http://acct-mgt..massopen.cloud
+
+  -- testing with basic authentication
+     python3 -m pytest amt_test*.py --amurl https://acct-mgt.massopen.cloud --basic "user:pass"
+"""
+
+import subprocess
+import re
+import time
+import json
+import pprint
+
+
+def get_pod_status(project, pod_name):
+    """This gets the specified pod status from openshift"""
+    result = subprocess.run(
+        ["oc", "-n", project, "-o", "json", "get", "pod", pod_name],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+        check=False,
+    )
+    if result.returncode == 0:
+        result_json = json.loads(result.stdout.decode("utf-8"))
+        print(result_json["status"]["phase"])
+        return result_json["status"]["phase"]
+    print("None")
+    return None
+
+
+#
+def wait_while(project, pod_name, statuses, time_out=300):
+    """
+    This basically waits while a pod is in a set of statuses and possibly times out
+    It is used to wait for openshift to do something with the pod, as in, deleting
+    or starting a pod.
+
+    pass in the following parameter:
+       project:  the namespace of the pod
+       pod_name: the name of the pod
+       statuses: array of statuses
+
+
+    """
+    time_left = time_out
+    time_interval = 5
+    time.sleep(time_interval)
+    status = get_pod_status(project, pod_name)
+    while status in statuses and time_left > 0:
+        time.sleep(time_interval)
+        time_left = time_left - time_interval
+        status = get_pod_status(project, pod_name)
+
+    if status in statuses:
+        return False
+    return True
+
+
+def wait_until_done(oc_cmd, finished_pattern, time_out=30, decrement=5):
+    """
+    This wait while an oc command is running, looking for a pattern that
+    indicates it is finished.
+    """
+    pattern = re.compile(finished_pattern)
+    done = False
+    oc_array = oc_cmd.split(" ")
+    matched_line = ""
+    while time_out > 0 and not done:
+        time.sleep(5)
+        time_out = time_out - decrement
+        result = subprocess.run(
+            oc_array,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            check=False,
+        )
+        line_list = result.stdout.decode("utf-8").split("\n")
+        for line in line_list:
+            if pattern.match(line):
+                matched_line = line
+                done = True
+    if pattern.match(matched_line):
+        return True
+    return False
+
+
+def compare_results(result, pattern):
```

Comment:
Do you really need to process the output line by line? Wouldn't it be sufficient to replace a call to `compare_result` with something like:

```
assert re.match(pattern, result.stdout.decode())
```

Or for simple strings:

```
assert pattern in result.stdout.decode()
```

Reply:
I would prefer to not assert things as that would cause the test script to halt.  I would prefer that it continues to completion as each one cleans up after itself.  I thought of writing general clean up code, but some things are easier to test as you are cleaning up.

Reply:
> I would prefer to not assert things as that would cause the test script to halt. I would prefer that it continues to completion as each test cleans up after itself while it is in the process of testing.

You can generally want tests to fail as rapidly as possible, but I understand the need to clean up resources.

You can have cleanup code that runs even if the test fails with an assert either by writing pytest fixtures that are context managers (see e.g. https://docs.pytest.org/en/6.2.x/fixture.html#yield-fixtures-recommended) or by using `try/finally` blocks.

Reply:
@rob-baron here's an example of a test suite that uses pytest fixtures to perform post-test cleanup, even if the test exits with an `assert` or an unexpected exception:

https://gist.github.com/larsks/99a8c7ef3d0ba0c1d57e024fb0f25de7

Reply:
@larsks, thank you for the example.  I wrote a similar one a while ago.  And yes, pytest, like every other testing framework provides a way to have setup code and cleanup code.  I haven't had a pressing need to do this as the set of 3 tests run though within 5 minutes - unlike other sets of tests that I have worked with, which took a couple of hours to run though.

I had initially used asserts as opposed to checks.  It turned out to be a painful experience to debug what was working and what was failing.  To be honest, it was mostly silly things like oc deciding to return an error code when it was able to connect to openshift and could not find a resource, such as either a user or a project - especially when I was testing for the non-existents of the user or project.

I agree with you in that most of the time you want to fail a test fast, and, to be honest.  Also I have considered adding  tests in such a manner that each test only tests one part of the interface.  This way each possible call to the micro server is tested in it's own individual test - ie, TestDeletionOfNonexistantProject would first check to see if the project it was deleting didn't exist using oc (oc would indicate success by passing back an error code).  Then check the delete error code and message.  finally confirming that the project still doesn't exist using oc.

Furthermore, testing individual parts of the interface does not neglect testing all of the parts of the interface.  Basically, I have opted for this as it is closer to how I expect this micro server is going to be used.  What I have written are integration tests and not unit tests.